gordon_janitor_gcp/gce_instance_authority.py

```diff
@@ -0,0 +1,145 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2018 Spotify AB
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+A GCEInstanceAuthority object retrieves a list of all instances in all
+projects that the ``keyfile`` gives it access to. It is meant to work
+in an asynchronous context and to be executed through the ``start()`` method.
+
+It is also meant to be easy to extend and customize so all methods except
+``start`` can be overridden as long as they follow the interfaces over the
+base methods. For example, if a different project filtering scheme is to be
+used, it's possible to inherit from GCEInstanceAuthority and override
+the ``get_all_projects`` method and ensure it returns a set of project IDs.
+
+To use:
+
+.. code-block:: pycon
+
+    >>> import asyncio
+    >>> rrset_channel = asyncio.Queue()
+    >>> authority = GCEInstanceAuthority(config, rrset_channel)
+    >>> await authority.start()
+    >>> msg = await.channel.get()
+    >>> print(msg)
+    {'zone': 'us-central1-b', 'resourceRecords': [...]}
+"""
+
+import aiohttp
+
+from gordon_janitor_gcp import auth
+from gordon_janitor_gcp import gce_clients
+from gordon_janitor_gcp import msg_generators
+
+
+class GCEInstanceAuthority:
+    """Gordon-Janitor plugin that gathers "state of truth" GCE instance data.
+
+    Args:
+        config (dict): Plugin-specific configuration.
+        rrset_channel (asyncio.Queue): channel to receive instance data
+            messages.
+    """
+    def __init__(self, config, rrset_channel=None, **kwargs):
+        self.config = config
+        self.keyfile = self.config['keyfile']
```

Comment:
Another unhandled `KeyError`. It should be the responsibility of the plugins to handle config, including erroring out if a required config item is needed, and should probably be done as early as possible - i.e. when plugins are loaded & initialized with the plugin-specific config. For example, the reconciler logic [errors out if missing a `keyfile` key](https://github.com/spotify/gordon-janitor-gcp/blob/develop/gordon_janitor_gcp/gdns_reconciler.py#L105-L111) as well as [a project key](https://github.com/spotify/gordon-janitor-gcp/blob/develop/gordon_janitor_gcp/gdns_reconciler.py#L119-L124).